src/coreclr/jit/abi.h

```diff
@@ -36,18 +36,134 @@ class ABIPassingSegment
 struct ABIPassingInformation
 {
     // The number of segments used to pass the value. Examples:
-    // - On x86, TYP_LONG can be passed in two registers, resulting in two
-    // register segments
     // - On SysV x64, structs can be passed in two registers, resulting in two
     // register segments
     // - On arm64/arm32, HFAs can be passed in up to four registers, giving
     // four register segments
     // - On arm32, structs can be split out over register and stack, giving
     // multiple register segments and a struct segment.
-    // - On Windows x64, all parameters always belong into one stack slot or register,
-    // and thus always have NumSegments == 1
+    // - On Windows x64, all parameters always fit into one stack slot or
+    // register, and thus always have NumSegments == 1
     unsigned           NumSegments = 0;
     ABIPassingSegment* Segments    = nullptr;
 
     bool IsSplitAcrossRegistersAndStack() const;
+
+    static ABIPassingInformation FromSegment(Compiler* comp, const ABIPassingSegment& segment);
+};
+
+class RegisterQueue
+{
+    const regNumberSmall* m_regs;
+    unsigned int          m_numRegs;
+    unsigned int          m_index = 0;
+
+public:
+    RegisterQueue(const regNumberSmall* regs, unsigned int numRegs) : m_regs(regs), m_numRegs(numRegs)
+    {
+    }
+
+    unsigned Count()
+    {
+        return m_numRegs - m_index;
+    }
+
+    regNumber Dequeue();
+    regNumber Peek();
+    void      Clear();
+};
+
+struct ClassifierInfo
+{
+    CorInfoCallConvExtension CallConv   = CorInfoCallConvExtension::Managed;
+    bool                     IsVarArgs  = false;
+    bool                     HasThis    = false;
+    bool                     HasRetBuff = false;
+};
+
+class X86Classifier
```

Comment:
What do you think about having target-specific classifiers in target-specific files (`targetXYZ.cpp`, which we already have but don't really use a lot)?

Reply:
I like that idea. They'll probably have to stay as being defined in `abi.h`, but I'll try moving the classification Implementation into those target specific files.